recv/Network/Socket/BufferPool/Recv.hs

```diff
@@ -69,9 +79,29 @@ tryRecvN init0 siz0 recv
                         go build' left'
 
 concatN :: Int -> [ByteString] -> ByteString
-concatN total bss0 = unsafeCreate total $ \ptr -> goCopy bss0 ptr
+-- Just because it's logical
+concatN _ [] = ""
+-- To avoid a copy if there's only one ByteString
+concatN _ [bs] = bs
+concatN total bss0 =
+    unsafeCreate total $ \ptr -> goCopy bss0 ptr
   where
     goCopy [] _ = return ()
     goCopy (bs : bss) ptr = do
         ptr' <- copy ptr bs
         goCopy bss ptr'
+
+_iorefRecv :: [ByteString] -> IO (IO ByteString)
+_iorefRecv ini = do
+    ref <- newIORef ini
+    return $ recv ref
+  where
+    recv ref = do
+        xxs <- readIORef ref
+        case xxs of
+            [] -> do
+                writeIORef ref $ error "closed"
+                return ""
+            x : xs -> do
```

Comment:
I know this is only used in a doctest, but technically, this might return an `empty ByteString` prematurely if one of the elements is empty.

```hs
    case xxs of
        [] -> do
            writeIORef ref $ error "closed"
            return ""
        x : xs -> do
            writeIORef ref xs
            if BS.null x
                -- if an element is empty, keep going
                then recv ref
                else return x
```

Reply:
A comment is added.